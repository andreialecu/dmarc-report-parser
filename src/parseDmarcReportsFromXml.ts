import { parseStringPromise } from "xml2js";
import { DmarcFeedbackSchema } from "./dmarc-feedback-schema";

/**
 * Parses DMARC reports from an array of XML strings or Buffers
 * @throws Error if the XML is not valid or does not match the DMARC schema
 */
export async function parseDmarcReportsFromXml(xmlFiles: (string | Buffer)[]) {
  const dmarcReports = (
    await Promise.all(xmlFiles.map((xml) => parseStringPromise(xml)))
  ).map((result) => DmarcFeedbackSchema.parse(result.feedback));

  // Analyze the DMARC reports and identify any failures
  const failures = dmarcReports.flatMap(({ record }) => {
    return record.filter((r) => {
      return r.row.policyEvaluated.disposition !== "none";
    });
  });

  const dkimFailures = dmarcReports.flatMap(({ record }) => {
    return record.filter((r) => {
      return r.row.policyEvaluated.dkim === "fail";
    });
  });

  const spfFailures = dmarcReports.flatMap(({ record }) => {
    return record.filter((r) => {
      return r.row.policyEvaluated.spf === "fail";
    });
  });

  return {
    totalReports: dmarcReports.length,
    totalFailures: failures.length,
    failures,
    dkimFailures,
    spfFailures,
    reports: dmarcReports,
  };
}
