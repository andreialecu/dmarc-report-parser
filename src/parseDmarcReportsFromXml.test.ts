import { readFileSync } from "fs";
import { join } from "path";
import { parseDmarcReportsFromXml } from "./parseDmarcReportsFromXml";

describe("parseDmarcReportsFromXml", () => {
  const validXml = readFileSync(
    join(__dirname, "./test/fixtures/xml/report-valid.xml"),
    "utf-8"
  );
  const quarantineXml = readFileSync(
    join(__dirname, "./test/fixtures/xml/report-quarantine.xml"),
    "utf-8"
  );
  const rejectXml = readFileSync(
    join(__dirname, "./test/fixtures/xml/report-reject.xml"),
    "utf-8"
  );
  const dkimFailXml = readFileSync(
    join(__dirname, "./test/fixtures/xml/report-dkim-fail.xml"),
    "utf-8"
  );
  const spfFailXml = readFileSync(
    join(__dirname, "./test/fixtures/xml/report-spf-fail.xml"),
    "utf-8"
  );

  const invalidXml = readFileSync(
    join(__dirname, "./test/fixtures/xml/report-invalid.xml"),
    "utf-8"
  );

  it("should parse a single valid DMARC report", async () => {
    const result = await parseDmarcReportsFromXml([validXml]);
    expect(result.totalReports).toBe(1);
    expect(result.reports.length).toBe(1);
    expect(result).toMatchSnapshot();
  });

  it("should parse multiple valid DMARC reports", async () => {
    const result = await parseDmarcReportsFromXml([validXml, validXml]);
    expect(result.totalReports).toBe(2);
    expect(result.reports.length).toBe(2);
  });

  it("should parse a DMARC report with a quarantine disposition", async () => {
    const result = await parseDmarcReportsFromXml([quarantineXml]);
    expect(result.totalReports).toBe(1);
    expect(result.reports.length).toBe(1);
    expect(result.failures).toHaveLength(1);
    expect(result).toMatchSnapshot();
  });

  it("should parse a DMARC report with a reject disposition", async () => {
    const result = await parseDmarcReportsFromXml([rejectXml]);
    expect(result.totalReports).toBe(1);
    expect(result.reports.length).toBe(1);
    expect(result.failures).toHaveLength(1);
    expect(result).toMatchSnapshot();
  });

  it("should parse a DMARC report that is passing but has a DKIM failure", async () => {
    const result = await parseDmarcReportsFromXml([dkimFailXml]);
    expect(result.totalReports).toBe(1);
    expect(result.reports.length).toBe(1);
    expect(result.failures).toHaveLength(0);
    expect(result.dkimFailures).toHaveLength(1);
    expect(result).toMatchSnapshot();
  });

  it("should parse a DMARC report with a SPF failure", async () => {
    const result = await parseDmarcReportsFromXml([spfFailXml]);
    expect(result.totalReports).toBe(1);
    expect(result.reports.length).toBe(1);
    expect(result.failures).toHaveLength(0);
    expect(result.spfFailures).toHaveLength(1);
    expect(result).toMatchSnapshot();
  });

  it("should throw an error when parsing an invalid DMARC report", async () => {
    await expect(parseDmarcReportsFromXml([invalidXml])).rejects.toThrow();
  });

  it("should throw an error when parsing multiple reports of which some are invalid", async () => {
    await expect(
      parseDmarcReportsFromXml([validXml, invalidXml])
    ).rejects.toThrow();
  });
});
