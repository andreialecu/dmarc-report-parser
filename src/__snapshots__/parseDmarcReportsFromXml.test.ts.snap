// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseDmarcReportsFromXml > should parse a DMARC report that is passing but has a DKIM failure 1`] = `
{
  "dkimFailures": [
    {
      "authResults": {
        "dkim": [
          {
            "domain": "bix-business.com",
            "humanResult": "",
            "result": "fail",
          },
        ],
        "spf": [
          {
            "domain": "bix-business.com",
            "result": "pass",
          },
        ],
      },
      "identifiers": {
        "headerFrom": "bix-business.com",
      },
      "row": {
        "count": 2,
        "policyEvaluated": {
          "disposition": "none",
          "dkim": "fail",
          "spf": "pass",
        },
        "sourceIp": "203.0.113.209",
      },
    },
  ],
  "failures": [],
  "reports": [
    {
      "policyPublished": {
        "adkim": "r",
        "aspf": "r",
        "domain": "bix-business.com",
        "p": "none",
        "pct": 100,
        "sp": "none",
      },
      "record": [
        {
          "authResults": {
            "dkim": [
              {
                "domain": "bix-business.com",
                "humanResult": "",
                "result": "fail",
              },
            ],
            "spf": [
              {
                "domain": "bix-business.com",
                "result": "pass",
              },
            ],
          },
          "identifiers": {
            "headerFrom": "bix-business.com",
          },
          "row": {
            "count": 2,
            "policyEvaluated": {
              "disposition": "none",
              "dkim": "fail",
              "spf": "pass",
            },
            "sourceIp": "203.0.113.209",
          },
        },
      ],
      "reportMetadata": {
        "dateRange": {
          "begin": 1335571200,
          "end": 1335657599,
        },
        "email": "noreply-dmarc-support@solarmora.com",
        "extraContactInfo": "http://solarmora.com/dmarc/support",
        "orgName": "solarmora.com",
        "reportId": "9391651994964116463",
      },
    },
  ],
  "spfFailures": [],
  "totalFailures": 0,
  "totalReports": 1,
}
`;

exports[`parseDmarcReportsFromXml > should parse a DMARC report with a SPF failure 1`] = `
{
  "dkimFailures": [],
  "failures": [],
  "reports": [
    {
      "policyPublished": {
        "adkim": "r",
        "aspf": "r",
        "domain": "bix-business.com",
        "p": "none",
        "pct": 100,
        "sp": "none",
      },
      "record": [
        {
          "authResults": {
            "dkim": [
              {
                "domain": "bix-business.com",
                "humanResult": "",
                "result": "pass",
              },
            ],
            "spf": [
              {
                "domain": "bix-business.com",
                "result": "fail",
              },
            ],
          },
          "identifiers": {
            "headerFrom": "bix-business.com",
          },
          "row": {
            "count": 2,
            "policyEvaluated": {
              "disposition": "none",
              "dkim": "pass",
              "spf": "fail",
            },
            "sourceIp": "203.0.113.209",
          },
        },
      ],
      "reportMetadata": {
        "dateRange": {
          "begin": 1335571200,
          "end": 1335657599,
        },
        "email": "noreply-dmarc-support@solarmora.com",
        "extraContactInfo": "http://solarmora.com/dmarc/support",
        "orgName": "solarmora.com",
        "reportId": "9391651994964116463",
      },
    },
  ],
  "spfFailures": [
    {
      "authResults": {
        "dkim": [
          {
            "domain": "bix-business.com",
            "humanResult": "",
            "result": "pass",
          },
        ],
        "spf": [
          {
            "domain": "bix-business.com",
            "result": "fail",
          },
        ],
      },
      "identifiers": {
        "headerFrom": "bix-business.com",
      },
      "row": {
        "count": 2,
        "policyEvaluated": {
          "disposition": "none",
          "dkim": "pass",
          "spf": "fail",
        },
        "sourceIp": "203.0.113.209",
      },
    },
  ],
  "totalFailures": 0,
  "totalReports": 1,
}
`;

exports[`parseDmarcReportsFromXml > should parse a DMARC report with a quarantine disposition 1`] = `
{
  "dkimFailures": [
    {
      "authResults": {
        "dkim": [
          {
            "domain": "bix-business.com",
            "humanResult": "",
            "result": "fail",
          },
        ],
        "spf": [
          {
            "domain": "bix-business.com",
            "result": "fail",
          },
        ],
      },
      "identifiers": {
        "headerFrom": "bix-business.com",
      },
      "row": {
        "count": 2,
        "policyEvaluated": {
          "disposition": "quarantine",
          "dkim": "fail",
          "spf": "fail",
        },
        "sourceIp": "203.0.113.209",
      },
    },
  ],
  "failures": [
    {
      "authResults": {
        "dkim": [
          {
            "domain": "bix-business.com",
            "humanResult": "",
            "result": "fail",
          },
        ],
        "spf": [
          {
            "domain": "bix-business.com",
            "result": "fail",
          },
        ],
      },
      "identifiers": {
        "headerFrom": "bix-business.com",
      },
      "row": {
        "count": 2,
        "policyEvaluated": {
          "disposition": "quarantine",
          "dkim": "fail",
          "spf": "fail",
        },
        "sourceIp": "203.0.113.209",
      },
    },
  ],
  "reports": [
    {
      "policyPublished": {
        "adkim": "r",
        "aspf": "r",
        "domain": "bix-business.com",
        "p": "quarantine",
        "pct": 100,
        "sp": "none",
      },
      "record": [
        {
          "authResults": {
            "dkim": [
              {
                "domain": "bix-business.com",
                "humanResult": "",
                "result": "fail",
              },
            ],
            "spf": [
              {
                "domain": "bix-business.com",
                "result": "fail",
              },
            ],
          },
          "identifiers": {
            "headerFrom": "bix-business.com",
          },
          "row": {
            "count": 2,
            "policyEvaluated": {
              "disposition": "quarantine",
              "dkim": "fail",
              "spf": "fail",
            },
            "sourceIp": "203.0.113.209",
          },
        },
      ],
      "reportMetadata": {
        "dateRange": {
          "begin": 1335571200,
          "end": 1335657599,
        },
        "email": "noreply-dmarc-support@solarmora.com",
        "extraContactInfo": "http://solarmora.com/dmarc/support",
        "orgName": "solarmora.com",
        "reportId": "9391651994964116463",
      },
    },
  ],
  "spfFailures": [
    {
      "authResults": {
        "dkim": [
          {
            "domain": "bix-business.com",
            "humanResult": "",
            "result": "fail",
          },
        ],
        "spf": [
          {
            "domain": "bix-business.com",
            "result": "fail",
          },
        ],
      },
      "identifiers": {
        "headerFrom": "bix-business.com",
      },
      "row": {
        "count": 2,
        "policyEvaluated": {
          "disposition": "quarantine",
          "dkim": "fail",
          "spf": "fail",
        },
        "sourceIp": "203.0.113.209",
      },
    },
  ],
  "totalFailures": 1,
  "totalReports": 1,
}
`;

exports[`parseDmarcReportsFromXml > should parse a DMARC report with a reject disposition 1`] = `
{
  "dkimFailures": [
    {
      "authResults": {
        "dkim": [
          {
            "domain": "bix-business.com",
            "humanResult": "",
            "result": "fail",
          },
        ],
        "spf": [
          {
            "domain": "bix-business.com",
            "result": "fail",
          },
        ],
      },
      "identifiers": {
        "headerFrom": "bix-business.com",
      },
      "row": {
        "count": 2,
        "policyEvaluated": {
          "disposition": "reject",
          "dkim": "fail",
          "spf": "fail",
        },
        "sourceIp": "203.0.113.209",
      },
    },
  ],
  "failures": [
    {
      "authResults": {
        "dkim": [
          {
            "domain": "bix-business.com",
            "humanResult": "",
            "result": "fail",
          },
        ],
        "spf": [
          {
            "domain": "bix-business.com",
            "result": "fail",
          },
        ],
      },
      "identifiers": {
        "headerFrom": "bix-business.com",
      },
      "row": {
        "count": 2,
        "policyEvaluated": {
          "disposition": "reject",
          "dkim": "fail",
          "spf": "fail",
        },
        "sourceIp": "203.0.113.209",
      },
    },
  ],
  "reports": [
    {
      "policyPublished": {
        "adkim": "r",
        "aspf": "r",
        "domain": "bix-business.com",
        "p": "reject",
        "pct": 100,
        "sp": "none",
      },
      "record": [
        {
          "authResults": {
            "dkim": [
              {
                "domain": "bix-business.com",
                "humanResult": "",
                "result": "fail",
              },
            ],
            "spf": [
              {
                "domain": "bix-business.com",
                "result": "fail",
              },
            ],
          },
          "identifiers": {
            "headerFrom": "bix-business.com",
          },
          "row": {
            "count": 2,
            "policyEvaluated": {
              "disposition": "reject",
              "dkim": "fail",
              "spf": "fail",
            },
            "sourceIp": "203.0.113.209",
          },
        },
      ],
      "reportMetadata": {
        "dateRange": {
          "begin": 1335571200,
          "end": 1335657599,
        },
        "email": "noreply-dmarc-support@solarmora.com",
        "extraContactInfo": "http://solarmora.com/dmarc/support",
        "orgName": "solarmora.com",
        "reportId": "9391651994964116463",
      },
    },
  ],
  "spfFailures": [
    {
      "authResults": {
        "dkim": [
          {
            "domain": "bix-business.com",
            "humanResult": "",
            "result": "fail",
          },
        ],
        "spf": [
          {
            "domain": "bix-business.com",
            "result": "fail",
          },
        ],
      },
      "identifiers": {
        "headerFrom": "bix-business.com",
      },
      "row": {
        "count": 2,
        "policyEvaluated": {
          "disposition": "reject",
          "dkim": "fail",
          "spf": "fail",
        },
        "sourceIp": "203.0.113.209",
      },
    },
  ],
  "totalFailures": 1,
  "totalReports": 1,
}
`;

exports[`parseDmarcReportsFromXml > should parse a single valid DMARC report 1`] = `
{
  "dkimFailures": [
    {
      "authResults": {
        "dkim": [
          {
            "domain": "bix-business.com",
            "humanResult": "",
            "result": "fail",
          },
        ],
        "spf": [
          {
            "domain": "bix-business.com",
            "result": "pass",
          },
        ],
      },
      "identifiers": {
        "headerFrom": "bix-business.com",
      },
      "row": {
        "count": 2,
        "policyEvaluated": {
          "disposition": "none",
          "dkim": "fail",
          "spf": "pass",
        },
        "sourceIp": "203.0.113.209",
      },
    },
  ],
  "failures": [],
  "reports": [
    {
      "policyPublished": {
        "adkim": "r",
        "aspf": "r",
        "domain": "bix-business.com",
        "p": "none",
        "pct": 100,
        "sp": "none",
      },
      "record": [
        {
          "authResults": {
            "dkim": [
              {
                "domain": "bix-business.com",
                "humanResult": "",
                "result": "fail",
              },
            ],
            "spf": [
              {
                "domain": "bix-business.com",
                "result": "pass",
              },
            ],
          },
          "identifiers": {
            "headerFrom": "bix-business.com",
          },
          "row": {
            "count": 2,
            "policyEvaluated": {
              "disposition": "none",
              "dkim": "fail",
              "spf": "pass",
            },
            "sourceIp": "203.0.113.209",
          },
        },
      ],
      "reportMetadata": {
        "dateRange": {
          "begin": 1335571200,
          "end": 1335657599,
        },
        "email": "noreply-dmarc-support@solarmora.com",
        "extraContactInfo": "http://solarmora.com/dmarc/support",
        "orgName": "solarmora.com",
        "reportId": "9391651994964116463",
      },
    },
  ],
  "spfFailures": [],
  "totalFailures": 0,
  "totalReports": 1,
}
`;
